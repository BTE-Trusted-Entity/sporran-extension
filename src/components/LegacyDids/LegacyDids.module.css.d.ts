export const __esModule: true;
export const overlay: string;
export const heading: string;
export const explanation: string;
export const cta: string;
export const listHeading: string;
export const list: string;
export const legacyIdentity: string;
export const close: string;
