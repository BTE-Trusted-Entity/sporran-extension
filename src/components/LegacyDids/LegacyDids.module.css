.overlay {
  composes: overlay from '../../components/Overlay/Overlay.module.css';
  max-height: calc(428rem / 16);
  text-align: center;
  padding: 1.5rem;
  overflow-y: auto;
}

.heading {
  composes: importantHeadline from '../../components/Typography/Typography.module.css';
  font-size: 0.875rem;
  padding-top: 4.5rem;
  background: url('../../components/images/exclamation.svg') no-repeat top
    center;
  margin-top: 0;
}

.explanation,
.cta {
  composes: smallText from '../../components/Typography/Typography.module.css';
  color: var(--color-grey-light);
  margin: 0 0 0.5rem;
}

.listHeading {
  composes: importantHeadline from '../../components/Typography/Typography.module.css';
  font-size: 0.875rem;
  margin: 0 0 0.25rem;
}

.list {
  list-style: none;
  padding: 0;
  margin: 0;
}

.legacyIdentity {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
  height: auto;
  font-size: 0.75rem;
  color: var(--color-orange);
  padding: 0.25rem 0;
}

.close {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
  margin-top: 0.5rem;
}
