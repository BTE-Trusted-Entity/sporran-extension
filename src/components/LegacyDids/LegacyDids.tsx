import { Modal } from 'react-dialog-polyfill';
import { browser } from 'webextension-polyfill-ts';
import { generatePath, Link } from 'react-router-dom';

import { useEffect } from 'react';

import * as styles from './LegacyDids.module.css';

import { paths } from '../../views/paths';

import { useLegacyDidIdentities } from '../../utilities/legacyDids/legacyDids';
import { useBooleanState } from '../../utilities/useBooleanState/useBooleanState';

export function LegacyDids(): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const legacyDidIdentities = useLegacyDidIdentities();

  const open = useBooleanState();

  useEffect(() => {
    open.set(Object.values(legacyDidIdentities).length > 0);
  }, [legacyDidIdentities, open]);

  if (!open.current) {
    return null;
  }

  return (
    <Modal open className={styles.overlay}>
      <h1 className={styles.heading}>{t('component_LegacyDids_heading')}</h1>

      <p className={styles.explanation}>
        {t('component_LegacyDids_explanation')}
      </p>
      <p className={styles.cta}>{t('component_LegacyDids_cta')}</p>

      <h3 className={styles.listHeading}>{t('component_LegacyDids_dids')}</h3>
      <ul className={styles.list}>
        {Object.values(legacyDidIdentities).map((identity) => (
          <li key={identity.address}>
            <Link
              to={generatePath(paths.identity.overview, {
                address: identity.address,
              })}
              onClick={open.off}
              className={styles.legacyIdentity}
            >
              {identity.name}
            </Link>
          </li>
        ))}
      </ul>

      <button className={styles.close} onClick={open.off}>
        {t('common_action_close')}
      </button>
    </Modal>
  );
}
