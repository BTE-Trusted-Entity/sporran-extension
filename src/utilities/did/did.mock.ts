import { isFullDid } from './did';

jest.mock('./did');

export function mockIsFullDid(boolean: boolean): void {
  jest.mocked(isFullDid).mockReturnValue(boolean);
}
