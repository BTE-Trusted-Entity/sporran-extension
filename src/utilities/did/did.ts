import {
  ConfigService,
  Did,
  DidDocument,
  DidEncryptionKey,
  DidUri,
  Utils,
} from '@kiltprotocol/sdk-js';

import { useAsyncValue } from '../useAsyncValue/useAsyncValue';

export function isFullDid(did: DidUri | undefined): boolean {
  if (!did) {
    // could be a legacy identity without DID
    return false;
  }
  return Did.parse(did).type === 'full';
}

export async function getFullDidDocument(did: DidUri): Promise<DidDocument> {
  const api = ConfigService.get('api');
  return Did.linkedInfoFromChain(await api.call.did.query(Did.toChain(did)))
    .document;
}

export function useFullDidDocument(did: DidUri): DidDocument | undefined {
  return useAsyncValue(getFullDidDocument, [did]);
}

export async function getDidDocument(did: DidUri): Promise<DidDocument> {
  return isFullDid(did)
    ? await getFullDidDocument(did)
    : Did.parseDocumentFromLightDid(did);
}

export function getDidEncryptionKey({
  keyAgreement,
}: DidDocument): DidEncryptionKey {
  if (!keyAgreement || !keyAgreement[0]) {
    throw new Error('encryptionKey is not defined somehow');
  }
  return keyAgreement[0];
}

export function parseDidUri(did: DidUri): ReturnType<typeof Did.parse> & {
  fullDid: DidUri;
} {
  const parsed = Did.parse(did);
  const fullDid = parsed.type === 'full' ? did : Did.getFullDidUri(did);

  return {
    ...parsed,
    fullDid,
  };
}

export async function needLegacyDidCrypto(
  did: DidUri | undefined,
): Promise<boolean> {
  if (!did) {
    // DID was deactivated, no action needed.
    return false;
  }
  if (!isFullDid(did)) {
    return false;
  }

  try {
    const encryptionKey = getDidEncryptionKey(await getDidDocument(did));
    return (
      Utils.Crypto.u8aToHex(encryptionKey.publicKey) ===
      '0xf2c90875e0630bd1700412341e5e9339a57d2fefdbba08de1cac8db5b4145f6e'
    );
  } catch {
    // getDidDocument might throw if the DID is not on-chain anymore (removed, another endpoint),
    // no legacy crypto needed in that case
    return false;
  }
}
