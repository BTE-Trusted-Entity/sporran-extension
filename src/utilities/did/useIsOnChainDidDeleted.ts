import { Did, DidUri } from '@kiltprotocol/sdk-js';

import { useAsyncValue } from '../useAsyncValue/useAsyncValue';

async function getIsOnChainDidDeleted(
  did: DidUri | undefined,
): Promise<boolean> {
  if (!did) {
    return false;
  }

  try {
    const resolved = await Did.resolve(did);
    return Boolean(
      resolved && resolved.metadata && resolved.metadata.deactivated,
    );
  } catch (error) {
    console.error(error, 'Could not get DID deletion status');
    return false;
  }
}

export function useIsOnChainDidDeleted(
  did: DidUri | undefined,
): boolean | undefined {
  return useAsyncValue(getIsOnChainDidDeleted, [did]);
}
