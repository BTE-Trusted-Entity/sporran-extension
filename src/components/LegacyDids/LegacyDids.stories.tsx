import { Meta } from '@storybook/react';

import {
  IdentitiesProviderMock,
  legacyIdentity,
} from '../../utilities/identities/IdentitiesProvider.mock';

import { LegacyDids } from './LegacyDids';

export default {
  title: 'Components/LegacyDids',
  component: LegacyDids,
} as Meta;

export function Template(): JSX.Element {
  return (
    <IdentitiesProviderMock
      identities={{ [legacyIdentity.address]: legacyIdentity }}
    >
      <LegacyDids />
    </IdentitiesProviderMock>
  );
}
