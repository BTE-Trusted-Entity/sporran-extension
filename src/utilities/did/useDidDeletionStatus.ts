import { IDidDetails } from '@kiltprotocol/types';
import { DidResolver } from '@kiltprotocol/did';

import { useAsyncValue } from '../useAsyncValue/useAsyncValue';

async function getDidDeletionStatus(did: IDidDetails['did']): Promise<boolean> {
  if (!did) {
    return false;
  }

  try {
    const resolved = await DidResolver.resolveDoc(did);
    return Boolean(
      resolved && resolved.metadata && resolved.metadata.deactivated,
    );
  } catch (error) {
    console.error(error);
    throw new Error('Could not get DID deletion status');
  }
}

export function useDidDeletionStatus(
  did: IDidDetails['did'],
): boolean | undefined {
  return useAsyncValue(getDidDeletionStatus, [did]);
}
