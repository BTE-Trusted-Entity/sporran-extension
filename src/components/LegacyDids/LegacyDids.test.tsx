import {
  mockDialogShowModal,
  render,
  waitForDialogUpdate,
} from '../../testing/testing';
import { needLegacyDidCrypto } from '../../utilities/did/did';

import { LegacyDids } from './LegacyDids';

describe('LegacyDids', () => {
  beforeEach(mockDialogShowModal);

  it('should render', async () => {
    jest.mocked(needLegacyDidCrypto).mockResolvedValue(true);

    const { container } = render(<LegacyDids />);
    await waitForDialogUpdate();
    expect(container).toMatchSnapshot();
  });
});
