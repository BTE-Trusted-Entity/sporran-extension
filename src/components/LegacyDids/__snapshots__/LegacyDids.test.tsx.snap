// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LegacyDids should render 1`] = `
<div>
  <dialog
    class="overlay"
  >
    <h1
      class="heading"
    >
      Important Update
    </h1>
    <p
      class="explanation"
    >
      We are updating DIDs for initial users, to reset the DID encryption feature for messaging in the future.
    </p>
    <p
      class="cta"
    >
      Please update your DIDs by initiating a transaction for each one that you set up. We have already refunded the transaction fee to you.
    </p>
    <h3
      class="listHeading"
    >
      DIDs that need updating:
    </h3>
    <ul
      class="list"
    >
      <li>
        <a
          class="legacyIdentity"
          href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1"
        >
          Light DID Identity
        </a>
      </li>
      <li>
        <a
          class="legacyIdentity"
          href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo"
        >
          Identity with on-chain DID and lots of coins
        </a>
      </li>
      <li>
        <a
          class="legacyIdentity"
          href="/identity/4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653"
        >
          Identity which has a very very very very long name and might not even fit on the screen
        </a>
      </li>
    </ul>
    <button
      class="close"
    >
      Close
    </button>
  </dialog>
</div>
`;
